smolyay/samples.py

```diff
@@ -0,0 +1,362 @@
+    the amount of points added by each level.
+    `points` describes the unique points added as the number of levels increase,
+    and the order of elements is reflective of the level each element first
+    appears in. Some point set are nested, meaning that the points in a previous
+    level reappear in following levels. However, `points` does not attempt to
+    capture this behavior and should be free of duplicates.
+
+    Parameters
+    ----------
+    max_level ; int
+        the maximum level the points are used for
+    """
+
+    def __init__(self, max_level):
+        super().__init__()
+        self.max_level = max_level
+        self._num_points = None
+        self._start_level = None
+        self._end_level = None
+
+    @property
+    def max_level(self):
+        """int: maximum level to compute points for."""
+        return self._max_level
+
+    @max_level.setter
+    def max_level(self, value):
+        self._max_level = value
+        self._valid_cache = True
+
+    @property
+    def num_points(self):
+        """list: number of points per level."""
+        if self._valid_cache:
+            self._create()
+        self._valid_cache = False
+        return self._num_points
+
+    @property
+    def start_level(self):
+        """list: the starting index of each level."""
+        if self._valid_cache:
+            self._create()
+        self._valid_cache = False
+        return self._start_level
+
+    @property
+    def end_level(self):
+        """list: the ending index of each level."""
+        if self._valid_cache:
+            self._create()
+        self._valid_cache = False
+        return self._end_level
+
+
+class ClenshawCurtisPointSet(UnidimensionalPointSet):
+    r"""Set of unidimensional points for Clenshaw Curtis sampling
+
+    The :attr:`points` for this interpolation scheme are the extrema of the
+    Chebyshev polynomials of the first kind on the domain :math:`[-1, 1]`:
+
+    .. math::
+
+        x_i^* = -\cos(\pi i/n), i = 0,...,n
+
+    For the special case :math:`n = 0`, there is only one point :math:`x_0^* = 0`.
+
+    Parameters
+    ----------
+    degree : int
+        degree of the Chebyshev polynomial of the first kind to get extrema from
+    """
+
+    def __init__(self, degree):
+        super().__init__()
+        self.degree = degree
+
+    @property
+    def domain(self):
+        """numpy.ndarray: Domain the sample points come from."""
+        return numpy.array([-1, 1])
+
+    @property
+    def degree(self):
+        """int: degree of polynomial to create points for."""
+        return self._degree
+
+    @degree.setter
+    def degree(self, value):
+        self._degree = value
+        self._valid_cache = True
+
+    def _create(self):
+        r"""Generating the points
+
+        Generating the extrema of a Cheybshev polynomial of the first kind at
+        a given degree.
+        """
+        if self.degree > 0:
+            points = -numpy.cos(
```

Comment:
Scale these by the domain. You can add a helper function somewhere for doing that transformation since it will be needed in most children.